### PR TITLE
certificate_manager: Check that template differs from current cert before rotation

### LINKED ISSUE #69471: Kubelet generates a new CSR on start even if it has a valid certificate on disk
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

I'm testing the new `--rotate-server-certificates` option in kubernetes 1.12 and it seems that kubelet always requests the signing of a new CSR and tries to rotate the certificates when it is started. This is happening even though there's a valid (brand new) certificate on disk issued by kube-controller-manager.

From what I can tell, this is happening because kubelet implements `getTemplate` and `lastRequest` is set after a successful certificate rotation, but is never persisted to disk. If kubelet is restarted, then the last requested CSR is nil.

https://github.com/kubernetes/client-go/blob/9c81604894197dbfe4424c19193068dc5211db9d/util/certificate/certificate_manager.go#L274-L283

**What you expected to happen**:

Ideally, kubelet should only request the signing of a new CSR if it doesn't already have a serving certificate or it's about to expire.

**How to reproduce it (as minimally and precisely as possible)**:

* Restart a kubelet instance that had it's serving certificate issued by kube-controller-manager
* Check `kubectl get csr` and see that it requested the signing of a new CSR

**Anything else we need to know?**:

kubelet is being run with `--rotate-server-certificates`. `--cert-dir`, `--tls-cert-file`, and `--tls-private-key-file` are all omitted to use the default values.

The TLS bootstrapping for client certificates is currently not being used--just the bootstrapping for serving certificates.

**Environment**:
- Kubernetes version: v1.12.0 - built from source at the `v1.12.0` tag (0ed33881dc4355495f623c6f22e7dd0b7632b7c0)
- Cloud provider or hardware configuration: Bare metal
- OS: CentOS 7.5.1804
- Kernel: `3.10.0-862.14.4.el7.x86_64`
- Install tools: Custom RPMs built from this git repo with no additional patches
- Others: `docker-1.13.1-75.git8633870.el7.centos.x86_64`

### PR DESCRIPTION
**Pull request by agunnerson-ibm**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

With the current behavior, when kubelet starts, a `templateChanged`
event is always fired off because it only checks if `getLastRequest`
matches `getTemplate`. The last request only exists in memory and thus
is initially `nil` and can't ever match the current template during
startup.

This causes kubelet to request the signing of a new CSR every time it's
restarted. This commit changes the behavior so that `templateChanged` is
only fired off if the currently template doesn't match both the current
certificate and the last template.

**Which issue(s) this PR fixes**:

Fixes #69471

**Release note**:

```release-note
When `--rotate-server-certificates` is enabled, kubelet will no longer request a new certificate on startup if the current certificate on disk is satisfactory.
```

### CODE REVIEW

**Comment by liggitt** [comment]
there's a gofmt error that needs fixing to make the bot happy:
```
-func (m* manager) certSatisfiesTemplate() bool {
+func (m *manager) certSatisfiesTemplate() bool {
 	m.certAccessLock.RLock()
 	defer m.certAccessLock.RUnlock()
 	if m.cert == nil {

Run ./hack/update-gofmt.sh
```

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager.go [hunk 1]
```diff
@@ -389,35 +389,30 @@ func (m *manager) rotateCerts() (bool, error) {
...
 	}
 
-	// Ensure the currently held certificate satisfies the requested subject CN and SANs
 	if template := m.getTemplate(); template != nil {
 		if template.Subject.CommonName != m.cert.Leaf.Subject.CommonName {
-			glog.V(2).Infof("Current certificate CN (%s) does not match requested CN (%s), rotating now", m.cert.Leaf.Subject.CommonName, template.Subject.CommonName)
-			return time.Now()
+			glog.V(2).Infof("Current certificate CN (%s) does not match requested CN (%s)", m.cert.Leaf.Subject.CommonName, template.Subject.CommonName)
+			return false
 		}
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager.go]: also check `sets.NewString(m.cert.Leaf.Subject.Organizations...).HasAll(template.Subject.Organizations...)`

**Comment by liggitt** [comment]
thanks for the PR, just a couple comments. exercising the logic in a unit test would be good as well

**Comment by agunnerson-ibm** [comment]
Thanks everyone for the comments! I've updated this PR to address them.

Changes:
* Fixed all gofmt errors
* Added check to ensure that the organizations in the x509 subject of the cert match the template
* Split `certSatisfiesTemplate` into `certSatisfiesTemplate` and `certSatisfiesTemplateLocked` and use the latter if `certAccessLock` is already locked
* Added unit tests for `certSatisfiesTemplate`

**Comment by agunnerson-ibm** [comment]
Fixed the lint issue reported by `pull-kubernetes-verify`.

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go [hunk 2]
```diff
@@ -212,6 +213,178 @@ func TestSetRotationDeadline(t *testing.T) {
...
+			}
+
+			m := manager{
+				cert:        tlsCert,
+				getTemplate: func() *x509.CertificateRequest { return tc.template },
+				usages:      []certificates.KeyUsage{},
+			}
+
+			if m.certSatisfiesTemplate() != tc.shouldSatisfy {
+				if tc.shouldSatisfy {
```
Comment (awly) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: Don't wrap errors, store `certSatisfiesTemplate` result in a var and do a simpler error print:
```
t.Errorf("cert: %+v, template: %+v, certSatisfiesTemplate returned %v, want %v", m.cert, tc.template, got, tc.shouldSatisfy)
```

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go [hunk 3]
```diff
@@ -212,6 +213,178 @@ func TestSetRotationDeadline(t *testing.T) {
...
+			if tc.cert != nil {
+				tlsCert = &tls.Certificate{
+					Leaf: tc.cert,
+				}
+			}
+
+			m := manager{
+				cert:        tlsCert,
+				getTemplate: func() *x509.CertificateRequest { return tc.template },
+				usages:      []certificates.KeyUsage{},
```
Comment (awly) [inline staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: Is this needed?

**Review by awly — COMMENTED** [review]
Two small nits, LGTM otherwise

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go (hunk 3 above)
Comment (agunnerson-ibm) [reply staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: Nope, I'll remove it.

File: staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go (hunk 2 above)
Comment (agunnerson-ibm) [reply staging/src/k8s.io/client-go/util/certificate/certificate_manager_test.go]: Thanks, I'll fix that.

**Comment by agunnerson-ibm** [comment]
The latest commit should address @awly's comments above.

**Pull request merged**